projects/plugins/jetpack/_inc/client/my-plan/my-plan-header/license.jsx

```diff
@@ -0,0 +1,77 @@
+/**
+ * External dependencies
+ */
+import React, { useState, useCallback } from 'react';
+import { __, _x } from '@wordpress/i18n';
+import { connect } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import TextInput from 'components/text-input';
+import {
+	successNotice as successNoticeAction,
+	errorNotice as errorNoticeAction,
+} from 'components/global-notices/state/notices/actions';
+import restApi from 'rest-api';
+
+const License = ( { errorNotice, successNotice } ) => {
+	const [ isSaving, setIsSaving ] = useState( false );
+	const [ licenseKeyText, setLicenseKeyText ] = useState( '' );
+
+	const handleInputChange = useCallback( event => {
+		setLicenseKeyText( event.target.value );
+	}, [] );
+
+	const saveJetpackLicense = useCallback( () => {
+		if ( ! licenseKeyText || isSaving ) {
+			return;
+		}
+
+		setIsSaving( true );
+
+		restApi
+			.updateLicenseKey( licenseKeyText )
+			.then( () => {
+				successNotice(
+					__(
+						'Jetpack license key added. It may take a minute for the license to be processed.',
+						'jetpack'
+					)
+				);
+
+				setIsSaving( false );
+				setLicenseKeyText( '' );
```

Comment:
Do you think we could modify a bit the UI so site owners can see when a license has been set? I'm afraid that if we don't, we'll cause some confusion. Even if we refactor / redesign that interface in the future, the version that we'll release in 9.6.0 will be available on some users' sites and may stay there if they don't update. I'd consequently prefer if it were as usable as possible.

I just created #19135 as a bit of a proposal you could maybe start from, or let me know what you think?

Reply:
I don't think this may be entirely possible to achieve. Right now the endpoint that Jetpack calls to get its available  plans only gets store subscriptions. There is no indication that says whether any subscriptions are linked to a license. There is ongoing work to eventually be able to do this but we are still at least a month away from that if not more.

The only thing we could possibly know right now is if the option has one or more licenses (what you are checking in #19135). These licenses may or may not be valid. Even if they are there is no way to tell if they have been used on some other site so we can't really consider what's in the option as the source of truth. @atanas-dev do you think that when the attach process fails we can remove that license from the option? So at least for now, only licenses that we know that _at some point_ were valid will remain in there and we have _something_ to show?

Having said that, this option can very easily get out of sync which I think will generate a lot more confusion than the current UI. For example, a partner/user puts a license in there. Then an HE or partner revokes the license. At this point the option is showing stale (and invalid) data and we have no way of updating it. Also, this option can hold an array of licenses which may look very strange as a list of comma-separated values and it may not be immediately obvious to a user that to add a new license they'd have to append it to the list.

I do understand that right now the process is less than ideal but this is just until the work in pbtFPC-1ek-p2 gets done. Exposure to this UI will be very very minimal.

Reply:
> I do understand that right now the process is less than ideal but this is just until the work in pbtFPC-1ek-p2 gets done. Exposure to this UI will be very very minimal.

I cannot say I'm a fan of delivering something that's not finished to customers, but I understand. I assume it wouldn't be an option to have no UI for now, and only instructions for customers to add a code snippet to their site to add a license via a filter, for example? A bit like what we do for affiliates and partner IDs today?

Reply:
> I assume it wouldn't be an option to have no UI for now, and only instructions for customers to add a code snippet to their site to add a license via a filter, for example? A bit like what we do for affiliates and partner IDs today?

This already exists and is in use for Jurassic Ninja sites as far as I understand. What we're aiming to provide is something basic that simplifies things even further.

> I cannot say I'm a fan of delivering something that's not finished to customers

Since this UI is hidden behind a filter, which is effectively a feature flag, I would lean towards us shipping and being clear with the agencies and hosting partners that are interested in this feature that it is not complete and that we may ask them to disable it if it causes a lot of confusion and/or support burden. Also, through this filter, we could work with agencies and hosting partners to conditionally allow the UI on certain version of Jetpack as well.